src/rpc/server/linux_io_uring.zig

```diff
@@ -0,0 +1,847 @@
+            const is_at_least = version.isAtLeast(min_version) orelse break :can_use .check;
+            break :can_use if (is_at_least) .yes else .no;
+        },
+        else => .no,
+    };
+
+    pub const InitError = std.posix.MMapError || error{
+        EntriesZero,
+        EntriesNotPowerOfTwo,
+
+        ParamsOutsideAccessibleAddressSpace,
+        ArgumentsInvalid,
+        ProcessFdQuotaExceeded,
+        SystemFdQuotaExceeded,
+        SystemResources,
+
+        PermissionDenied,
+        SystemOutdated,
+    };
+
+    // NOTE(ink): constructing the return type as `E!?T`, where `E` and `T` are resolved
+    // separately seems to help ZLS with understanding the types involved better, which is
+    // why I've done it like that here. If ZLS gets smarter in the future, you could probably
+    // inline this into a single branch in the return type expression.
+    const InitErrOrEmpty = if (can_use == .no) error{} else InitError;
+    const InitResultOrNoreturn = if (can_use == .no) noreturn else LinuxIoUring;
+    pub fn init() InitErrOrEmpty!?InitResultOrNoreturn {
+        const need_runtime_check = switch (can_use) {
+            .no => return null,
+            .yes => false,
+            .check => true,
+        };
+
+        var io_uring = IoUring.init(4096, 0) catch |err| return switch (err) {
+            error.SystemOutdated,
+            error.PermissionDenied,
+            => |e| if (!need_runtime_check) e else return null,
+            else => |e| e,
+        };
+        errdefer io_uring.deinit();
+
+        return .{
+            .io_uring = io_uring,
+            .multishot_accept_submitted = false,
+            .pending_cqes_count = 0,
+            .pending_cqes_buf = undefined,
+        };
+    }
+
+    pub fn deinit(self: *LinuxIoUring) void {
+        self.io_uring.deinit();
+    }
+
+    pub const AcceptAndServeConnectionsError =
+        IouGetSqeError ||
+        IouSubmitError ||
+        ConsumeOurCqeError ||
+        std.mem.Allocator.Error;
+
+    pub fn acceptAndServeConnections(
+        self: *LinuxIoUring,
+        server_ctx: *server.Context,
+    ) AcceptAndServeConnectionsError!void {
+        if (!self.multishot_accept_submitted) {
+            self.multishot_accept_submitted = true;
+            errdefer self.multishot_accept_submitted = false;
+            _ = self.io_uring.accept_multishot(
+                @bitCast(Entry.ACCEPT),
+                server_ctx.tcp.stream.handle,
+                null,
+                null,
+                std.os.linux.SOCK.CLOEXEC,
+            ) catch |err| return switch (err) {
+                error.SubmissionQueueFull => |e| {
+                    server_ctx.logger.err().log(
+                        "Under normal circumstances the accept_multishot would be" ++
+                            " the first SQE to be queued, but somehow the queue was full.",
+                    );
+                    return e;
+                },
```

Comment:
is this worth being a panic?

Reply:
Hmm, I'm a bit indifferent, is there a specific reason you'd want it to be a panic vs a propagated error? Mind you, this isn't necessarily a one-time thing, `self.multishot_accept_submitted` can be reset if the accept_multishot is terminated (see line 184).

Reply:
The submissions queue being full would occur if theres many IO operations started. Could the RPC server reach that high, given enough accepted connections doing IO? If so, to continue serving it should probably submit() what's currently in the submission queue and try enqueueing the operation again (could maybe make this a wrapper fn)

Reply:
Under normal circumstances, it shouldn't occur, however it's not impossible that there is some subtle control flow oversight which could lead to that situation; thinking about it more, we probably would want it to panic in such a circumstance.

Reply:
Discussed on slack, resolution: 1242b40725d54ee352f8f47cb860851b790bfe82